src/collections.rs

```
@@ -0,0 +1,236 @@
+pub struct BlockingDeque<T> {
+    queue: Mutex<VecDeque<T>>,
+    control: (Sender<()>, Receiver<()>),
+}
+
+impl<T> BlockingDeque<T> {
+    /// Creates a new queue.
+    ///
+    /// The created queue has space to hold at most `cap` records at a time.
+    ///
+    /// # Panics
+    ///
+    /// Capacity must be a positive number. If `cap` is zero, this function will panic.
+    pub fn new(cap: usize) -> Self {
+        let queue = Mutex::new(VecDeque::new());
+        let control = async_std::channel::bounded(cap);
+        Self { queue, control }
+    }
+
+    /// Appends an element to the back of the deque.
+    ///
+    /// If the queue is full, this method waits until there is space for a value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use fluvio_future::collections::BlockingDeque;
+    /// fluvio_future::task::run(async {
+    ///     let queue = BlockingDeque::new(2);
+    ///     assert!(queue.is_empty());
+    ///     queue.push_back(1).await;
+    ///     queue.push_back(1).await;
+    ///     assert_eq!(queue.len(), 2);
+    /// });
+    /// ```
+    pub async fn push_back(&self, value: T) {
+        let _ = self.control.0.send(()).await; // drop the error because we own both sides
+        self.queue.lock().await.push_back(value);
+    }
+
+    /// Removes the first element and returns it, or `None` if the deque is
+    /// empty.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use fluvio_future::collections::BlockingDeque;
+    /// fluvio_future::task::run(async {
+    ///     let queue = BlockingDeque::new(1);
+    ///     assert_eq!(queue.pop_front().await, None);
+    ///     queue.push_back(1).await;
+    ///     assert_eq!(queue.pop_front().await, Some(1));
+    /// });
+    /// ```
+    pub async fn pop_front(&self) -> Option<T> {
+        let front = self.queue.lock().await.pop_front()?;
+        if let Err(err) = self.control.1.try_recv() {
+            error!(
```

why ignore the error here?